Charlesmchen/one system contacts fetch at a time

charlesmchen opened the pull request:
PTAL @michaelkirk

michaelkirk commented on Signal/src/contact/SystemContactsFetcher.swift:
```diff
@@ -434,15 +452,21 @@ class SystemContactsFetcher: NSObject {
         DispatchQueue.global().async {
 
-            var fetchedContacts: [Contact]?
+            guard self.tryToAcquireContactFetchLock() else {
```
Seems like there's a race condition where we could lose important contact edits this way. If I launch the app, then while my contacts are still loading, I edit someone in my address book, I might not get that edit pulled into Signal until *next* contact pull.

Maybe that's an edge case that we can reasonably ignore, but we could also use a serial queue instead.

WDYT @charlesmchen

charlesmchen replied:
Yup, that's the definite tradeoff here.  It's hard because this is a major perf win on iPhone 5C if you have a lot of contacts.

Maybe we can have two types of fetch requests: soft and hard.  Hard are always done, serially.  Soft can be de-duplicated.

michaelkirk commented on Signal/src/contact/SystemContactsFetcher.swift:
```diff
@@ -434,15 +452,21 @@ class SystemContactsFetcher: NSObject {
+                self.releaseContactFetchLock()
                 return
             }
+            self.releaseContactFetchLock()
```
maybe a good use for a swift `defer` block.

charlesmchen replied:
Done.

charlesmchen commented:
PTAL @michaelkirk 

I have:

* Renamed `ignoreDebounce` to `alwaysNotify`.
* Added `shouldDebounce` so that when system contacts change we can ensure that we always do a fetch.
* With the introduction of ``shouldDebounce` we can once again more than one fetch at a time, and thus race conditions but I'm not sure this is worth addressing.
* I wonder if we couldn't simplify by modifying `fetchIfAlreadyAuthorized` to noop if `systemContactsHaveBeenRequestedAtLeastOnce` like `requestOnce`.  We could then move fetches onto a serial queue, remove debounce logic. WDYT?

michaelkirk commented on Signal/src/contact/SystemContactsFetcher.swift:
```diff
@@ -417,25 +419,55 @@ class SystemContactsFetcher: NSObject {
+                    return
+                }
+            }
+            defer {
```
Can you put the defer in the previous `if shouldDebounce` block to put the "teardown" code as logically close to the "setup" code as possible.

charlesmchen replied:
No, because then the `defer` executes when you leave the condition block.

michaelkirk replied:
I am surprised and appalled.

michaelkirk commented:
> I wonder if we couldn't simplify by modifying fetchIfAlreadyAuthorized to noop if systemContactsHaveBeenRequestedAtLeastOnce like requestOnce. We could then move fetches onto a serial queue, remove debounce logic. WDYT?

I'm also open to this idea. Adding a "override" flag with a default value is expedient, but it gets harder to understand what the method does. Especially when there are multiple override flags with default values.

charlesmchen commented:
PTAL @michaelkirk Reworked this PR yet again, as discussed.
